<a href="https://colab.research.google.com/github/akanksha-cherukapalli/Simulation-and-modeling---CS6647XTIA/blob/main/Homework5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
Activity_state = {
    0: 'rest',
    1: 'sweeties',
    2: 'exercises'
}

state = Activity_state
print(state)


{0: 'rest', 1: 'sweeties', 2: 'exercises'}


The transition matrix ( for the given state diagram) is as follows:

In [8]:
import numpy as np

transition_matrix = np.array([[0.2, 0.2, 0.6],
                             [0.2, 0.1, 0.7],
                             [0.1, 0.3, 0.6]])

print(transition_matrix)


[[0.2 0.2 0.6]
 [0.2 0.1 0.7]
 [0.1 0.3 0.6]]


The random walk can be:

In [10]:
n = 15
starting_state = 0
previous_state = starting_state

# Performing the random walk
for i in range(n):
    current_state = np.random.choice([0, 1, 2], p=transition_matrix[previous_state])
    print(Activity_state[current_state], "---->", end="")
    # Update the previous state
    previous_state = current_state

rest ---->exercises ---->rest ---->exercises ---->exercises ---->exercises ---->rest ---->rest ---->rest ---->sweeties ---->exercises ---->rest ---->sweeties ---->exercises ---->sweeties ---->

The starting state is Rest. Calculate the probability of possible activity on the 15th day.
What about the state and probability after 10000 days?

In [11]:
import numpy as np

# Transition matrix
transition_matrix = np.array([[0.2, 0.2, 0.6],
                             [0.2, 0.1, 0.7],
                             [0.1, 0.3, 0.6]])

# Starting state vector
start_state = np.array([1, 0, 0])  # Rest is the starting state

# Calculating probability after 15 days
prob_15_days = start_state @ np.linalg.matrix_power(transition_matrix, 15)

# Calculating probability after 10,000 days
prob_10000_days = start_state @ np.linalg.matrix_power(transition_matrix, 10000)

# Printing the respective probabilities
print("Probability of activities on the 15th day:")
print("Rest:", prob_15_days[0])
print("Sweeties:", prob_15_days[1])
print("Exercises:", prob_15_days[2])

# Printing the respective probabilities
print("\nProbability of activities after 10,000 days:")
print("Rest:", prob_10000_days[0])
print("Sweeties:", prob_10000_days[1])
print("Exercises:", prob_10000_days[2])


Probability of activities on the 15th day:
Rest: 0.13761467889913795
Sweeties: 0.23853211009159786
Exercises: 0.6238532110092636

Probability of activities after 10,000 days:
Rest: 0.13761467889903567
Sweeties: 0.2385321100916618
Exercises: 0.6238532110089617


3. What do you observe from the above two?

Upon observation, we can notice that as the number of days increases, the probabilities of being in each state tend to stabilize. After a long duration, such as 10,000 days in this case, the probabilities converge to certain values. It indicates that the system reaches a steady distribution. Notably, the probabilities after 10,000 days are very similar to those after 15 days, suggesting that the distribution has reached equilibrium. As the number of steps or trials increases, the probabilities of different states become closer to each other. With a low number of trials, the model may yield higher probabilities for rest and sweets. However, as the number of trials increases, we observe that the probabilities settle around 13.76% for rest, 23.85% for sweets, and 62.39% for exercises.